src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,286 @@
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
+
+        replaceFunc(question);
+        replaceFunc(hints);
+        replaceFunc(particle);
+
+        // handle the case where the replaceString contains one or more copies
+        // of searchString
+        // TODO(kevinb7) this is quite right
+        var searchResultCount = this.getSearchResultCount(searchString);
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ question, hints, json: particle, searchIndex: 0 });
+    },
+
+    handleReplace() {
+        var replaceString = this.state.replaceString;
+        var { searchIndex, searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(searchString, "g");
+        var replaced = false;
+        var globalIndex = 0;
+
+        var replaceFunc = function(obj) {
+            if (!replaced && obj) {
+                if (Array.isArray(obj)) {
+                    obj.forEach(replaceFunc);
+                } else {
+                    // excludes occurrences within widget references
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        // make sure this match isn't inside a widget reference
+                        if (indices.indexOf(offset) !== -1) {
+                            if (!replaced && globalIndex === searchIndex) {
+                                replaced = true;
+                                globalIndex++;
+                                return replaceString;
+                            } else {
+                                globalIndex++;
+                                return match;
+                            }
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
+
+        replaceFunc(question);
+        replaceFunc(hints);
+        replaceFunc(particle);
+
+        var searchResultCount = this.state.searchResultCount;
+
+        if (replaceString.indexOf(searchString) === -1) {
+            // normal case: replaceString doesn't contain searchString so we
+            // can decrement searchResultCount because we just replaced one 
+            // occurence of searchString
+            searchResultCount --;
+        } else {
+            // edge case: replaceString contains one or more instances of
+            // searchString which means the searchResultCount will either 
+            // stay the same if the number of instances is one or increase
+            // if it's more than one
+            var matches = replaceString.match(regex);
+            if (matches) {
+                searchResultCount += matches.length - 1;
+            }
+        }
```

Comment:
Can we unify these into something like:

```javascript
var matches = replaceString.match(regex);
var matchCount = matches ? matches.length : 0
searchResultCount = this.state.searchResultCount + matchCount - 1;
```

A comment about what we're doing here + why would be nice too :)